servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -174,57 +206,63 @@ private DnsDiscoveryObserver newDiscoveryObserver(final String address) {
+        // any pending scheduled tasks. SrvInactiveCombinerOperator is used to filter the aggregated collection of
+        // inactive events if necessary.
+        return recoverWithInactiveEvents(new SrvRecordPublisher(serviceName, discoveryObserver), true)
+                .flatMapConcatIterable(identity())
+                .flatMapMerge(srvEvent -> {
                 assertInEventloop();
                 if (srvEvent.isAvailable()) {
-                    final ARecordPublisher aPublisher =
-                            new ARecordPublisher(false, srvEvent.address().hostName(), discoveryObserver);
-                    final ARecordPublisher prevAPublisher = aRecordMap.putIfAbsent(srvEvent.address(), aPublisher);
-                    if (prevAPublisher != null) {
-                        return failed(new IllegalStateException("Only 1 A* record per SRV record is supported. " +
-                                srvEvent.address() + " corresponding to SRV name " + serviceName +
-                                " had a pre-existing A* record:" + prevAPublisher.name +
-                                " when new A* record arrived: " + aPublisher.name));
-                    }
+                    return defer(() -> {
+                        final ARecordPublisher aPublisher =
+                                new ARecordPublisher(srvEvent.address().hostName(), discoveryObserver);
+                        final ARecordPublisher prevAPublisher = aRecordMap.putIfAbsent(srvEvent.address(), aPublisher);
+                        if (prevAPublisher != null) {
+                            return newDuplicateSrv(serviceName, srvEvent.address().toString(), prevAPublisher.name);
+                        }
 
-                    return srvARecordPubToSingle(aPublisher, srvEvent, serviceName);
+                        Publisher<? extends Collection<ServiceDiscovererEvent<InetAddress>>> returnPub =
+                                recoverWithInactiveEvents(aPublisher, false);
+                        return srvFilterDuplicateEvents ?
+                                srvFilterDups(returnPub, availableAddresses, srvEvent.address().port()) :
+                                returnPub.map(events -> mapEventList(events, inetAddress ->
+                                        new InetSocketAddress(inetAddress, srvEvent.address().port())));
+                    }).retryWhen((i, cause) -> {
+                        assertInEventloop();
+                        // If this error is because the SRV entry was detected as inactive, then propagate the error and
+                        // don't retry. Otherwise this is a resolution exception (e.g. UnknownHostException), and retry.
+                        return cause == SrvAddressRemovedException.DNS_SRV_ADDR_REMOVED ||
+                                aRecordMap.remove(srvEvent.address()) == null ?
+                                Completable.failed(cause) : srvHostNameRepeater.apply(i);
```

Comment:
There are two levels of lookups. This code propagates failures at the first level (e.g. the DNS server returns nxdomain for the SRV query, netty times it out, etc.). However this code retries on the A record Publishers. if a failure on a single A record publisher is propagated this may lead to not processing complete results, and kill the entire SRV lookup (impacting other A records that are resolving fine) for scenario that maybe temporary (e.g. spurious errors due to DNS reconfiguration and distributed caches, A record TTL is lower than SRV record TTL and gets invalidated first, etc.). Consider the below example, if the error is processed first we wouldn't process subsequent results that maybe successful.

```
// SRV records
SRV (hostname1.com, 9999)
SRV (hostname2.com, 9999)
SRV (hostname3.com, 9999)

// DNS records
<nxdomain/timeout for hostname2>
A (hostname3.com, 1.2.3.4)
A (hostname1.com, 1.2.3.5)
```